### PR TITLE
openjdk: enable bootstrapping on ARM

### PR DESCRIPTION
**Pull request by lopsided98**
###### Motivation for this change

I started out with the goal of adding support for 32-bit ARM to `adoptopenjdk-bin`, but then quickly realized that this binary package could be used to bootstrap the OpenJDK build on these platforms. This led to me also making a lot of changes to clean up the OpenJDK packages.

###### Things done

This PR makes `openjdk8` and `openjdk11` use `adoptopenjdk-bin` for bootstrapping. [AdoptOpenJDK](https://adoptopenjdk.net/) is a well-known organization that provides prebuilt OpenJDK binaries for many architectures. Their build scripts are open source and their infrastructure is publicly visible, which I believe makes Nix's OpenJDK bootstrapping process more transparent and trustworthy than the status quo, which uses a bootstrapping tarball of undocumented origin. From a technical standpoint, AdoptOpenJDK binaries make it easier to bootstrap OpenJDK on more platforms.

While I was working on this, I tried to clean up parts of the packages that I believe are no longer necessary or could be done better, but in doing so I there is the possibility that I broke something. Are there any particularly problematic packages I should test?

###### Testing

|  | `x86_64-linux` | `i686-linux` | `aarch64-linux` | `armv6l-linux` | `armv7l-linux` |
|----------------------|----------------|--------------|-----------------|----------------|----------------|
| `openjdk8` | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :no_entry_sign: | :no_entry_sign: |
| `openjdk8_headless` | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :no_entry_sign: | :no_entry_sign: |
| `openjdk11` | :heavy_check_mark: | :x:<sup>1</sup> | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| `openjdk11_headless` | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| `bazel` | :heavy_check_mark: | | :heavy_check_mark: | | | 
| `jmeter` | :heavy_check_mark: | | | | |
| `jetbrains.idea-community`<sup>2</sup> | :heavy_check_mark: | | | |
| `javaPackages` | :heavy_check_mark: |

:heavy_check_mark: = builds and runs
:x: = fails to build
:no_entry_sign: = not supported
blank = not tested

<sup>1</sup>`openjfx11` (and therefore `openjdk11`) fails to build on `i686-linux` for non-obvious reasons, even without this PR, and I did not try to fix it.

<sup>2</sup>Overridden to use OpenJDK 11 built from source

I do not have the ability to test on macOS, but I did not really touch anything Darwin related.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of binary files (usually in `./result/bin/`) (I didn't test every single binary)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by lopsided98** [comment]
Some other notes:

`openjfx11` does not support ARM, but Debian has some patches that add support. I spent a little time trying to integrate them, but gave up because my aarch64 machines don't have enough RAM and there were strange errors on 32-bit ARM.

OpenJDK 8 does not directly support ARM, but there is an upstream branch (which is used by AdoptOpenJDK for their builds) which supports `aarch64`. Again, Debian has some patches that make it work on 32-bit ARM, but I did not try to integrate them.

**Comment by volth** [comment]
Cool.
`graalvm8` (https://github.com/NixOS/nixpkgs/pull/61475) might work on `aarch64-linux` with your fixes.
It does not work only because of lack of working `openjdk8`

**Comment by lheckemann** [comment]
>gave up because my aarch64 machines don't have enough RAM

Are you aware of the community builder? https://github.com/nix-community/aarch64-build-box

**Comment by lopsided98** [comment]
Yes, I thought about trying the community builder, but I was getting kind of tired of working on this PR and I didn't really like the idea of cluttering nixpkgs with a bunch of patches for a feature that likely gets little  use.

File: pkgs/development/compilers/adoptopenjdk-bin/generate-sources.py
Comment (timokau) [inline pkgs/development/compilers/adoptopenjdk-bin/generate-sources.py]: Is that a debugging leftover?

File: pkgs/development/compilers/adoptopenjdk-bin/generate-sources.py
Comment (timokau) [inline pkgs/development/compilers/adoptopenjdk-bin/generate-sources.py]: I'd keep the comment mentioning "jdk-11.0.1+13" as an example name, makes it easier to understand what is going on here.

File: pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix
Comment (timokau) [inline pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix]: Do we not need this backwards compatibility any more?

File: pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix
Comment (timokau) [inline pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix]: Why is that not needed anymore?

**Review by timokau — APPROVED** [review]
LGTM after some superficial review. I don't know anything about jdk bootstrapping, so I can't greenlight this on my own in good conscience. But this looks like a great improvement, thank you for this!

Some nitpicks included.

**Comment by timokau** [comment]
CC @taku0 who init'ed adoptopenjdk and may be able to provide a more useful review. Also @bennofs, @tomfitzhenry, @wkennington, @NeQuissimus, @thoughtpolice as a semi-random sample of people intereseted in (adopt-)openjdk according to git shortlog. Hopefully we'll be able to keep this from bitrotting too badly :)

File: pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix
Comment (lopsided98) [reply pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix]: `autoPatchelfHook` does this now

**Comment by taku0** [comment]
Running a [simple AWT application](https://github.com/taku0/JavaSample) results in `NullPointerException`:
```
Caused by: java.lang.NullPointerException
        at java.desktop/sun.awt.FontConfiguration.getVersion(Unknown Source)
        at java.desktop/sun.awt.FontConfiguration.readFontConfigFile(Unknown Source)
        at java.desktop/sun.awt.FontConfiguration.init(Unknown Source)
        at java.desktop/sun.awt.X11FontManager.createFontConfiguration(Unknown Source)
        at java.desktop/sun.font.SunFontManager$2.run(Unknown Source)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.desktop/sun.font.SunFontManager.<init>(Unknown Source)
        at java.desktop/sun.awt.FcFontManager.<init>(Unknown Source)
        at java.desktop/sun.awt.X11FontManager.<init>(Unknown Source)
        ... 73 more
```

May or may not related to https://github.com/AdoptOpenJDK/openjdk-build/issues/693.

Reverting `autoPatchelf` to manual `patchelf` and adding `fontconfig` to the `buildInput` fixes this (I don't know why).
https://gist.github.com/taku0/74fe5aba9e9230b363dc8a19fc7c07f1

**Comment by lopsided98** [comment]
I fixed the fontconfig issue. There are a few other libraries that could use a similar fix (e.g. gtk3), but they shouldn't cause any crashes, just missing features such as native look and feel. Now that we have normal OpenJDK builds for all platforms that AdoptOpenJDK supports, very few users will need to use AdoptOpenJDK, so these features will probably not be missed.

I have tested with the example @taku0 posted, as well with some JavaFX examples I found here: https://github.com/callicoder/javafx-examples

File: pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix
Comment (lopsided98) [reply pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix]: I couldn't figure out what the backwards compatibility concern was here, since `openjdk11` does not have this symlink. Also, users are unlikely to use `adoptopenjdk` directly anymore, so I felt that backwards compatibility was less important.

Comment (timokau) [reply pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix]: Maybe @taku0 who added the line initially has an opinion?

**Comment by grahamc** [comment]
@GrahamcOfBorg eval

**Comment by timokau** [comment]
@lopsided98 would you mind rebasing this on current master? I don't know why ofBorg complains about a chromium trace, I can only assume it was an issue with the version of master this PR was branched off of.

File: pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix
Comment (taku0) [reply pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix]: Oracle JDK and other JDKs had `jre` folder and some packages depend on it.  I forgot which packages depend on it.

Comment (lopsided98) [reply pkgs/development/compilers/adoptopenjdk-bin/jdk-linux-base.nix]: If `openjdk11` doesn't have it, and no one's complaining, then I don't think `adoptopenjdk` needs it either.

**Comment by lopsided98** [comment]
I rebased it, but it didn't fix anything. I think ofborg merges the PR into master before testing, so it shouldn't matter what commit it is based on.

**Comment by timokau** [comment]
Okay looks like this line is causing the trace:
https://github.com/NixOS/nixpkgs/blob/07cf130686d98e05166dfdffe10dcac64379f415/pkgs/applications/networking/browsers/chromium/common.nix#L153
Which complains because the current oldest chromium version in nixpkgs is 76, so there is no point in keeping that old patch.

What I don't understand is why the trace only appears on this PR. It should pop up every time you eval chromium on a non-clang machine (e.g. x86 linux). What would make more sense is the patch below that, which is aarch64 specific. Aarch64 chromium probably wasn't eval'd by ofBorg before because the jdk was unfree and the aarch64 specific patch is also for an outdated version. But that is not the one the trace complains about.

@lopsided98 maybe just remove both outdated patches?

(CC chromium maintainers @bendlas @ivan)

**Comment by volth** [comment]
@timokau 
`protobuf` `python2.pkgs.jinja2` `python2.pkgs.ply` deps should be safe to remove as well (76+ does not need them)

**Comment by timokau** [comment]
I would argue that other chromium changes don't need to be included in this PR, removing the patches was only necessary to appease ofBorg. Any other cleanup can be done in followup PRs.

Any objections to merging this as-is? @taku0 I'd especially appreciate a final comment from you, since you did a lot of the prior work and had some comments that are (I think) resolved.

**Comment by ivan** [comment]
I am a nixpkgs chromium maintainer and that chromium change is fine with me.

**Comment by timokau** [comment]
I just noticed this targets master. It should probably target staging instead, both for the relatively high rebuild count and the additional testing.

**Comment by taku0** [comment]
AdoptOpenJDK seems to be fine.

File: pkgs/development/compilers/openjdk/openjfx/11.nix
Comment (abbradar) [inline pkgs/development/compilers/openjdk/openjfx/11.nix]: I remember faintly that for some reason `disallowedReferences` didn't work for me when I wrote this. Then again, I don't see _why_ should they fail and I failed to properly document the reason. If it builds it should work so let's keep it that way.

**Review by abbradar — APPROVED** [review]

**Comment by timokau** [comment]
I have rebased on staging, I hope you don't mind. Only conflict was with 396e3a6b3cf11578b1005fd6ad0513d6edaba46b.

Sorry @ code-owners, I don't know if there's a way to change the target branch without notifying everyone.

**Comment by timokau** [comment]
I've skimmed over the changes one more time, everything seems good. ofBorg approves as well. I'll do some builds on the aarch64 community builder as a sanity check, then merge.

File: pkgs/development/libraries/libdrm/default.nix
Comment (matthewbauer) [inline pkgs/development/libraries/libdrm/default.nix]: Since this is an optional dependency, we should have a flag that is something like ```withValgrind```. Also the meta.available check is not so good because it also takes into account allowBroken. Instead, we should use ```lib.any (lib.meta.platformMatch stdenv.hostPlatform) valgrind-light.meta.platforms```

File: pkgs/development/compilers/openjdk/openjfx/12.nix
Comment (matthewbauer) [inline pkgs/development/compilers/openjdk/openjfx/12.nix]: This could be a little bit smarter if you took ```lib.attrNames outputHashes```. It should still work for platforms, we just don't have hashes added for them.

**Comment by volth** [comment]
@GrahamcOfBorg build jvmci8
@GrahamcOfBorg build graalvm8

**Review by timokau — APPROVED** [review]
Sorry for the delay. I believe this is good to go now, as soon as @matthewbauer and @volth agree. @matthewbauer does have valid points, but I'd rather just merge this now instead of going through more iterations. Big changes like this tend to hang around forever if you want them to be absolutely perfect in the first try.

**Comment by lopsided98** [comment]
I'm working on addressing @matthewbauer's concerns right now

File: pkgs/development/libraries/libdrm/default.nix
Comment (lopsided98) [reply pkgs/development/libraries/libdrm/default.nix]: That check does not work because `valgrind` uses `meta.badPlatforms`.

**Comment by timokau** [comment]
@volth for what it's worth, I built jvmci8 on x86 locally. Not sure if I tested it on aarch64. I suspect ofBorg will time out.

File: pkgs/development/compilers/openjdk/openjfx/12.nix
Comment (lopsided98) [reply pkgs/development/compilers/openjdk/openjfx/12.nix]: I'm not sure if things have gotten better in OpenJFX 12, but OpenJFX 11 requires more than just new hashes to build on other platforms. I'm not sure I like this approach anyway because it makes the code less clear for little benefit.

**Comment by lopsided98** [comment]
I haven't really addressed the `valgrind-light` availability check, because I don't know that the proper solution is. Other than that I think this is ready.

**Comment by timokau** [comment]
@matthewbauer @volth objections to merge?

**Comment by volth** [comment]
> @matthewbauer @volth objections to merge?

LGTM if `graalvm8` builds (I have not tested and ofborg times out)

**Review by matthewbauer — APPROVED** [review]

**Comment by timokau** [comment]
graalvm builds on x86 (while eating a TON of ram, wow). Thanks again @lopsided98 for taking on this PR.

**Comment by FRidh** [comment]
This broke the tarball job on `staging`. Please fix because it would block the next `staging` -> `staging-next` merge.

**Comment by timokau** [comment]
Apparently for some reason `openjdk11.meta.license` cannot be found (no attribute `license`). Not sure why, but a hotfix would probably be to specify the license in `jfx11` explicitly. I don't currently have time to make a PR, will you take this on @lopsided98?

**Comment by lopsided98** [comment]
`openjdk11` on macOS is a Zulu JDK build () which does not have a license attribute. It's probably less error prone to just give OpenJFX its own `license` attribute, but I'll add the license to the macOS JDK as well.

On a related note, it would probably make sense to use AdoptOpenJDK instead of Zulu on macOS, but I don't have any macOS machines so I can't do it myself.

**Comment by lopsided98** [comment]
See #67839

**Comment by timokau** [comment]
@grahamc shouldn't ofBorg have caught this? Its essentially an eval failure on mac, isn't it?

**Comment by vcunat** [comment]
Any idea why this MR introduced this release-blocking breakage?  https://github.com/NixOS/nixpkgs/issues/68307

**Comment by fgaz** [comment]
I think this PR broke all gradle-based packages, specifically commit de5e65ad8054276a7c43b603ea491ba65db81547 (found by bisecting).

I am getting this error:

```
> Task :PD-classes:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':PD-classes:compileJava'.
> Could not find tools.jar. Please check that /nix/store/6l52g0dy4v2wrxjwrphcsl7wzf3brb9w-openjdk-8u212-ga-jre/lib/openjdk/jre contains a valid JDK installation.
```

For some reason gradle seems to use the jre instead of the jdk regardless of the `java` parameter.

**Comment by lopsided98** [comment]
@fgaz #68442 should fix it

File: pkgs/development/compilers/openjdk/8.nix
Comment (Ma27) [inline pkgs/development/compilers/openjdk/8.nix]: The JRE inputs need to be propagated to jdk as well to ensure that the `setJavaClassPath` hook is used.
This breaks for instance scalafmt[1] and scalafix[2].

* https://hydra.nixos.org/build/100896633
* https://hydra.nixos.org/build/100895668

related to ZHF #68361

**Pull request merged**
